TrueCacheAPI.md

```diff
@@ -0,0 +1,236 @@
+It then compares it with the Hibernate session start time to decide which came first.
+If the transaction that calls `CachedDomainDataAccess.putFromLoad` is older than the invalidation, invalidation wins.
+Otherwise, `CachedDomainDataAccess.putFromLoad` is allowed to store data into the cache.
+Over the years Hibernate users have come to rely on such behaviour, but such maintaining such strictness has a cost.  
+
+
+## Read-Write Cache
+
+A read-write cache supports a wider range of situations compared to a read-only cache:
+
+It handles updates which act on both the cached value and the value at source.
+It also handles removals which act both the cached value and the value at source.
+
+On top of that, the get function can compare the cache valued and the value fetched from the source.
+This is important in a read-write cache because the versions of the data in the cache and the data source might be different.
+This is not the case in a read-only cache. 
+
+```java
+interface ReadWriteCache<K, V> {
+
+  /**
+   * Checks if the cache contains a value associated with the key.
+   * If the value is present, it's returned immediately wrapped in a CompletionStage.
+   * 
+   * If the value is not present, the function is called asynchronously to fetch the value from an unknown source.
+   * In the presence of concurrent updates, a value might have been cached and updated by the time the fetcher returns.
+   * To handle such situations the user can provide a comparator function to help resolve discrepancies:
+   * 
+   * If a value exists in the cache when the fetcher completes, 
+   * the comparator is called with the fetched value as first parameter,
+   * and the cached value as second parameter.
+   * 
+   * If the comparator returns 0 or a negative number,
+   * it means the fetched value is the same or older than the cached value,
+   * in which case the fetched value is not cached and the cached value is returned wrapped in a CompletionStage.
+   * 
+   * If the comparator returns a positive number,
+   * the fetched value is newer than the cached value,
+   * in which case the fetched value is cached and returned wrapped in a CompletionStage.
+   * 
+   * @param key to look up in the cache
+   * @param fetcher called if not value is associated with the key in the cache
+   * @param comparator called if a cached value exists when the fetcher returns 
+   * @return a future containing the cached value or null if not values exist 
+   */
+  CompletionStage<V> get(K key, Function<K, V> fetcher, Comparator<V> comparator);
+
+  /**
+   * Remove entry from the cache and remove it from source by calling the remover function.
+   * 
+   * @param key to delete from cache and source
+   * @param remover function that deletes data from source
+   * @return a future that is completed when key has been removed from both cache and source  
+   */
+  CompletionStage<Void> remove(K key, Consumer<K> remover);
```

Comment:
The point of the `Consumer<K>` is that it defines how the key is removed in the source of truth (e.g. a database). Just like the `get()`, we want the cache to have some control over this so that it can easily know when the source of truth has completed. 

The problem with your suggestion is that the cache cannot know when the source of truth remove has completed.

The code above is blocking, so to be fully non-blocking it should look more like this:

```
CompletionStage<Void> remove(K key, Function<K, CompletionStage<Void>> remover);
```